File: python/Q56_merge_intervals.py
#
# @lc app=leetcode id=56 lang=python3
#
# [56] Merge Intervals
#

# @lc code=start
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        all_times: dict[int, int] = dict()

        for start, end in intervals:
            if start not in all_times:
                all_times[start] = 0

            if end not in all_times:
                all_times[end] = 0

            all_times[start] += 1
            all_times[end] -= 1

        processing: int = 0
        current_start: int | None = None

        result: list[list[int]] = []

        for t in sorted(all_times.keys()):
            if current_start is None:
                current_start = t

            processing += all_times[t]

            if processing == 0:
                result.append([current_start, t])
                current_start = None

        return result


# @lc code=end

File: rust/src/leetcode/Q56_merge_intervals.rs
/*
 * @lc app=leetcode id=56 lang=rust
 *
 * [56] Merge Intervals
 */

// @lc code=start
impl Solution {
    pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut intervals: Vec<Vec<i32>> = intervals;
        intervals.sort();

        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut start: i32 = intervals[0][0];
        let mut end: i32 = intervals[0][1];

        for i in 0..intervals.len() {
            end = end.max(intervals[i][1]);

            if i != intervals.len() - 1 && intervals[i + 1][0] <= end {
                continue;
            }

            result.push(vec![start, end]);

            if i != intervals.len() - 1 {
                start = intervals[i + 1][0];
                end = intervals[i + 1][1];
            }
        }

        result
    }
}
// @lc code=end
